File: graphr/src/graphr/processing/Agent.java
package graphr.processing;

import graphr.graph.Edge;


/**
 * Parent class for all graph's agents. User is expected to extend this class and implemented given methods.
 *
 */
public abstract class Agent {

	protected VertexProcessingFacade v;
	
	public void setVertexProcessingFacade(VertexProcessingFacade v) {
		this.v = v;
	}
	
	public abstract Agent getCopy();
	
	public abstract void setUsedEdge(Edge<?, ?> e);
	
	/**
	 * Called before simulation starts to do some initialization. Here comes user's code.
	 */
	public void runBefore(){}
	
	/**
	 * Called after simulation stops to do some clean up. Here comes user's code.
	 */
	public void runAfter(){}

	/**
	 * Perform computation on given vertex and decide where to move in the next step
	 * @param vertex On which vertex to perform computation
	 * @return null if agent halts otherwise where to move next
	 */
	public abstract void runStep();
	
}

File: graphrs/graphrs/processing/Agent.swift
//
//  Agent.swift
//  graphrs
//
//  Created by Marcus Sellmann on 25.03.16.
//  Copyright © 2016 CITBDA. All rights reserved.
//

import Foundation

/**
 * Parent class for all graph's agents. User is expected to extend this class and implemented given methods.
 */
protocol Agent {
    var v: VertexProcessingFacade { get set }
    
    func getCopy() -> Agent
    func setUsedEdge(e: Edge)
    
    /**
     * Called before simulation starts to do some initialization. Here comes user's code.
     */
    func runBefore()
    
    /**
     * Called after simulation stops to do some clean up. Here comes user's code.
     */
    func runAfter()
    
    /**
     * Perform computation on given vertex and decide where to move in the next step
     * @param vertex On which vertex to perform computation
     * @return null if agent halts otherwise where to move next
     */
    func runStep()
    
}